src/routes/publish.js

```diff
@@ -42,29 +45,78 @@ export function publishRoutes() {
 
     const hasherTransform = new HasherPassThrough();
     const lengthChecker = new LengthCheckerPassThrough(MAX_BODY_LENGTH);
+    // The payload should "look like" a json object.
+    const jsonChecker = new CheapJsonChecker();
+    const gunzipStream = new GunzipWrapper();
     const concatener = new Concatenator();
 
     // The "pipeline" utility conveniently destroys all streams when there's an
     // error. However the HTTP request isn't part of the pipeline because we
     // don't want to destroy it. Indeed if it's destroyed, the socket is closed
     // and we can't send a nice error to the caller.
-    const pipeline = util.promisify(Stream.pipeline);
+    const pipeline = promisify(Stream.pipeline);
+    const finished = promisify(Stream.finished);
+
+    // We create 2 interconnected pipelines:
+    //
+    //     request
+    //        |
+    //  lengthChecker --- gunzipStream --- jsonChecker
+    //        |
+    // hasherTransform
+    //        |
+    //   concatener
+    //
+    // These streams' possible errors are handled partly by nodejs' pipeline
+    // function, and some special error handlers to forward errors between
+    // pipelines.
+
+    // 1. First and main pipeline
     const pipelinePromise = pipeline(
       lengthChecker,
       hasherTransform,
       concatener
     );
 
+    // 2. "Gunzip and check" json pipeline
+    // This pipeline will stop quite early, once we know this looks like a json.
+    // When it ends we'll just unpipe.
+    // We don't use the `pipeline` utility because with it we get some
+    // "premature close" errors when the length checker errors out.
```

Comment:
Frankly I don't understand why we get these errors here but not in the previous pipeline... I spent too much time understanding all kind of errors and I'm a bit fed up with this, so I didn't want to invest more time now that everything works fine with the current implementation... I'm a bit frustrated that I don't understand this, so maybe I'll come back to it later, but I want to move on :-D

Reply:
Works for me, it's a bit unfortunate the streams API is so poorly designed.